privval/grpc/server_test.go

```diff
@@ -0,0 +1,193 @@
+
+	testCases := []struct {
+		name       string
+		pv         types.PrivValidator
+		have, want *types.Vote
+		err        bool
+	}{
+		{name: "valid", pv: types.NewMockPV(), have: &types.Vote{
+			Type:             tmproto.PrecommitType,
+			Height:           1,
+			Round:            2,
+			BlockID:          types.BlockID{Hash: hash, PartSetHeader: types.PartSetHeader{Hash: hash, Total: 2}},
+			Timestamp:        ts,
+			ValidatorAddress: valAddr,
+			ValidatorIndex:   1,
+		}, want: &types.Vote{
+			Type:             tmproto.PrecommitType,
+			Height:           1,
+			Round:            2,
+			BlockID:          types.BlockID{Hash: hash, PartSetHeader: types.PartSetHeader{Hash: hash, Total: 2}},
+			Timestamp:        ts,
+			ValidatorAddress: valAddr,
+			ValidatorIndex:   1,
+		},
+			err: false},
+		{name: "invalid vote", pv: types.NewErroringMockPV(), have: &types.Vote{
+			Type:             tmproto.PrecommitType,
+			Height:           1,
+			Round:            2,
+			BlockID:          types.BlockID{Hash: hash, PartSetHeader: types.PartSetHeader{Hash: hash, Total: 2}},
+			Timestamp:        ts,
+			ValidatorAddress: valAddr,
+			ValidatorIndex:   1,
+			Signature:        []byte("signed"),
+		}, want: &types.Vote{
+			Type:             tmproto.PrecommitType,
+			Height:           1,
+			Round:            2,
+			BlockID:          types.BlockID{Hash: hash, PartSetHeader: types.PartSetHeader{Hash: hash, Total: 2}},
+			Timestamp:        ts,
+			ValidatorAddress: valAddr,
+			ValidatorIndex:   1,
+			Signature:        []byte("signed"),
+		},
+			err: true},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			s := tmgrpc.SignerServer{
+				ChainID: ChainID,
+				PrivVal: tc.pv,
+			}
+			req := &privvalproto.SignVoteRequest{ChainId: s.ChainID, Vote: tc.have.ToProto()}
+			resp, err := s.SignVote(context.Background(), req)
+			if tc.err {
+				require.Error(t, err)
+			} else {
+				pbVote := tc.want.ToProto()
+
+				require.NoError(t, tc.pv.SignVote(s.ChainID, pbVote))
+				assert.Equal(t, pbVote.Signature, resp.Vote.Signature)
+			}
+		})
+	}
+}
+
+func TestSignProposal(t *testing.T) {
+
+	ts := time.Now()
+	hash := tmrand.Bytes(tmhash.Size)
+
+	testCases := []struct {
+		name       string
+		pv         types.PrivValidator
+		have, want *types.Proposal
+		err        bool
+	}{
+		{name: "valid", pv: types.NewMockPV(), have: &types.Proposal{
+			Type:      tmproto.ProposalType,
+			Height:    1,
+			Round:     2,
+			POLRound:  2,
+			BlockID:   types.BlockID{Hash: hash, PartSetHeader: types.PartSetHeader{Hash: hash, Total: 2}},
+			Timestamp: ts,
+		}, want: &types.Proposal{
+			Type:      tmproto.ProposalType,
+			Height:    1,
+			Round:     2,
+			POLRound:  2,
+			BlockID:   types.BlockID{Hash: hash, PartSetHeader: types.PartSetHeader{Hash: hash, Total: 2}},
+			Timestamp: ts,
+		},
+			err: false},
+		{name: "invalid proposal", pv: types.NewErroringMockPV(), have: &types.Proposal{
```

Comment:
It's an invalid proposal because it already contains a signature right?

Reply:
I thought at first but this is erroring because the MockPV errors on all calls.